### PR TITLE
Move backoff to per-JSONClient instead of per-POST

### PR DESCRIPTION
**Pull request by rolandshoemaker**
Makes POST backoff global in `JSONClient` instead of per `PostAndParseWithRetry`. If a `Retry-After` header is returned override any existing backoff that is set.

Hoists most of the backoff logic out of `PostAndParseWithRetry` and into top-level methods. I think overall this makes the logic a bit more easy to understand.

I've updated the existing retry/backoff test but have been scratching my head about the best way to test the concurrent blocking behavior.

Fixes google/certificate-transparency#1324.

### CODE REVIEW

**Comment by rolandshoemaker** [comment]
Test failures appear to be related to gRPC/etcd incompatibilities?

**Comment by daviddrysdale** [comment]
(Closed and re-opened to trigger a new Travis build, in the hope that #76 might have fixed things)

File: jsonclient/client.go [hunk 1]
```diff
@@ -41,6 +44,69 @@ type JSONClient struct {
 	httpClient *http.Client          // used to interact with the server via HTTP
 	Verifier   *ct.SignatureVerifier // nil for no verification (e.g. no public key available)
 	logger     Logger                // interface to use for logging warnings and errors
+
+	multiplier int64
+	until      *time.Time
```
Comment (daviddrysdale) [inline jsonclient/client.go]: Might be clearer renamed to `notBefore` (but maybe I've looked at too many certificates...)

File: jsonclient/client.go [hunk 2]
```diff
@@ -41,6 +44,69 @@ type JSONClient struct {
...
+	}
+}
+
+func (c *JSONClient) backoff(ctx context.Context) error {
+	c.mu.RLock()
+	if c.until == nil {
+		c.mu.RUnlock()
+		return nil
+	}
+	until := time.Until(*c.until)
```
Comment (daviddrysdale) [inline jsonclient/client.go]: `until` -> `notBefore` might be clearer again.

File: jsonclient/client.go [hunk 3]
```diff
@@ -41,6 +44,69 @@ type JSONClient struct {
...
+const (
+	maxBackoff = time.Second * 128
+	jitter     = 250
+)
+
+func (c *JSONClient) setBackoff(override *time.Duration) time.Duration {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	if c.until != nil {
+		return time.Until(*c.until)
```
Comment (daviddrysdale) [inline jsonclient/client.go]: Shouldn't this take `override` into account?  If we're already retrying and so have a `c.notBefore` of (say) 2s away, we should still respect a subsequent `Retry-After`

File: jsonclient/client.go [hunk 4]
```diff
@@ -41,6 +44,69 @@ type JSONClient struct {
...
+	}
+	var wait time.Duration
+	if override != nil {
+		wait = *override
+	} else {
+		c.multiplier++
```
Comment (daviddrysdale) [inline jsonclient/client.go]: Having no limit for this multiplier feels worrying, particularly for any long-lived client instances now that backoff state is persistent.  A couple of hours of backoffs pegged at 128s  will get the multiplier up to ~55, which will take a long time to get back down to 1.

File: jsonclient/client.go [hunk 5]
```diff
@@ -41,6 +44,69 @@ type JSONClient struct {
...
+			wait = maxBackoff
+		}
+	}
+	until := time.Now().Add(wait)
+	c.until = &until
+	go func() {
+		time.Sleep(wait)
+		c.mu.Lock()
+		defer c.mu.Unlock()
+		c.until = nil
```
Comment (daviddrysdale) [inline jsonclient/client.go]: Hmm, this relies on nothing else having changed `c.until` in the intervening interval, which I don't think is safe -- at the point when this goroutine gets kicked off, say with a 2s wailt, another goroutine might be blocked at the top of a `setBackoff(60s)` which will change `c.until` to 60s ... but `c.until` will get cleared 2s later.

Might be safer to just have `backoff()` cope with `notBefore` / `until` being in the past.

File: jsonclient/client.go [hunk 6]
```diff
@@ -41,6 +44,69 @@ type JSONClient struct {
...
 	logger     Logger                // interface to use for logging warnings and errors
+
+	multiplier int64
+	until      *time.Time
+	mu         sync.RWMutex
+}
+
+const (
+	maxBackoff = time.Second * 128
+	jitter     = 250
```
[the remark this answers is not in the document]
Comment (AlCutter) [reply jsonclient/client.go]: How about `jitter = 250 * time.Millisecond` and then at ~L351 something like `leeway := jitter * (test.failCount+1)`?

*#drive-by-codereview* :car:

**Comment by rolandshoemaker** [comment]
(Sorry for the lag)

Took a stab at refactoring the backoff logic into its own struct with the `JSONClient` using an interface so we can factor out the actual wall clock tests by using a mock `backoff` object. If this looks good I'll start re-working the tests to take advantage of this and add separate tests for the backoff specific code. Also probably the existing `backoff.backoff` method should be moved to the `JSONClient` and replaced with a simple `backoff.notBefore` method that returns the current `notBefore` time so that we don't need a fake clock abstraction for testing...

**Comment by rolandshoemaker** [comment]
Hurray, the `jsonclient` tests have gone from ~30s to <1s with this.

File: jsonclient/backoff.go [hunk 7]
```diff
@@ -0,0 +1,62 @@
+package jsonclient
```
Comment (daviddrysdale) [inline jsonclient/backoff.go]: The new files are going to need the Apache license header.

File: jsonclient/backoff_test.go [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+	// Test that the interval increases as expected
+	for i := uint(0); i < maxMultiplier; i++ {
+		n := time.Now()
+		interval := b.set(nil)
+		if interval != time.Second*(1<<i) {
+			t.Fatalf("backoff.set() returned an unexpected duration. wanted: %s, got: %s", time.Second*(1<<i), interval)
+		}
+		expected := n.Add(interval)
+		if !fuzzyTimeEquals(expected, b.notBefore, testLeeway) {
+			t.Fatalf("backoff.notBefore is not expected time. wanted: %s (+/- 10µs), got: %s", expected, b.notBefore)
```
Comment (daviddrysdale) [inline jsonclient/backoff_test.go]: This [looks to be failing](https://travis-ci.org/google/certificate-transparency-go/jobs/285813715) under `--race`...

File: jsonclient/backoff.go [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+	}
+}
+
+func (b *backoff) until() time.Time {
+	b.mu.RLock()
+	defer b.mu.RUnlock()
+	if b.notBefore.Before(time.Now()) {
+		return b.notBefore
+	}
+	return b.notBefore.Add(time.Millisecond * time.Duration(rand.Intn(int(maxJitter.Seconds()*1000))))
```
[the remark this answers is not in the document]
Comment (rolandshoemaker) [reply jsonclient/backoff.go]: I think that makes sense.

**Comment by daviddrysdale** [comment]
This has a knock-on effect on the timings for the `client/` tests, so I think we need something like #98 to go in before I merge.  (Aside: Travis seems happy though, which confuses me -- any idea why?)

**Comment by rolandshoemaker** [comment]
> This has a knock-on effect on the timings for the client/ tests, so I think we need something like #98 to go in before I merge. (Aside: Travis seems happy though, which confuses me -- any idea why?)

Ah, I had been using `scripts/presubmit.sh` so I hadn't noticed there were other tests that test retry/backoff behavior. Things like `TestAddChainRetries` get skipped in that case (and on Travis) because `go test --short ...` is used. For the full tests to continue working #98 will definitely need to go in first.

That said it seems like `TestAddChainRetries` could just be nuked since it's actually testing `jsonclient` functionality which is already fully covered.

**Comment by rolandshoemaker** [comment]
@daviddrysdale thanks sticking with me on this! Should hopefully lessen the impact on logs when they aren't doing too well in the future.

**Comment by daviddrysdale** [comment]
Thanks for persisting -- having fewer foot-guns out there is surely a good idea!

**Pull request merged**